Nouns 🤝 i18n

bcjgit opened the pull request:
# TLDR
This PR does the following:

1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
2. Add UI (mobile and desktop) to allow users to switch between Languages
3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
4. Add logic to detect user's system / browser locale (again, based off Uniswap) 
5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)

## Testing Done
1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome) 
3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic

solimander commented on packages/nouns-webapp/src/components/NavLocaleSwitcher/index.tsx:
```diff
@@ -0,0 +1,172 @@
+import LanguageSelectionModal from '../LanguageSelectionModal';
+import { setLocale } from '../../i18n/setLocale';
+import { Trans } from '@lingui/macro';
+import navWalletClasses from '../NavWallet/NavWallet.module.css';
```
Should rename this if used by both the nav wallet and language picker

solimander commented:
Language button + dropdown isn't aware of screen width changes. If you load at a small screen size and then widen, it still behaves as if it's on mobile (and vice versa).



e.g.

<img width="697" alt="desktop" src="https://user-images.githubusercontent.com/85371573/166080147-72998eff-3eea-4cd5-8a70-7610854f5dc7.png">

<img width="489" alt="mobile" src="https://user-images.githubusercontent.com/85371573/166080139-bf1e7cbf-cd80-45f4-940f-77bdd2bce2f1.png">

solimander commented on packages/nouns-webapp/src/i18n/setLocale.ts:
```diff
@@ -0,0 +1,16 @@
+export const setLocale = (locale: string) => {
+  if (localStorage.getItem('lang') !== locale) {
+    localStorage.setItem('lang', locale);
+    window.location.reload();
```
What's the purpose of the reload? Shouldn't be required to do a full page refresh on locale change.

bcjgit replied:
Could be missing something, but I found that the date localization didn't work until after a reload. Less importantly, this was also to fix issues with a few locale specific UI changes I made (spacing etc.) that look at `useActiveLocale` to trigger.

solimander replied:
I'll take a look at the date localization issue, but it'd be a much nicer UX if we didn't need to do a full-page refresh.

solimander replied:
Enabling `forceRenderOnLocaleChange` should solve the same issue and provide a cleaner UX. It will re-render the UI without requiring API data to be re-fetched.

bcjgit replied:
I had that enabled (I think) but still required a full reload to get the dates to change ... will also look into this. Fully ack that I could just be being stupid on this...

solimander commented on packages/nouns-webapp/package.json:
```diff
@@ -66,7 +71,11 @@
+    "i18n:extract": "lingui extract",
+    "i18n:compile": "lingui compile",
+    "i18n:pseudo": "lingui extract --locale pseudo && lingui compile"
+
```
Nit: Remove unused line

solimander commented on packages/nouns-webapp/src/components/Bid/index.tsx:
```diff
@@ -162,27 +167,35 @@ const Bid: React.FC<{
-          title: 'Transaction Failed',
-          message: placeBidState.errorMessage ? placeBidState.errorMessage : 'Please try again.',
+          title: <Trans>Transaction Failed</Trans>,
+          message: placeBidState.errorMessage ? (
+            placeBidState.errorMessage
+          ) : (
+            <Trans>Please try again.</Trans>
+          ),
```
I know this was already in place, but can make these more readable by using optional chaining and or operator.

e.g.

```ts
placeBidState?.errorMessage || <Trans>Please try again.</Trans>,
```

solimander commented on packages/nouns-webapp/src/components/Bid/index.tsx:
```diff
@@ -276,7 +298,7 @@ const Bid: React.FC<{
             <Col lg={12} className={classes.voteForNextNounBtnWrapper}>
               <Button className={classes.bidBtnAuctionEnded} onClick={fomoNounsBtnOnClickHandler}>
-                Vote for the next Noun ⌐◧-◧
+                <Trans>Vote for the next Noun ⌐◧-◧</Trans>
```
Exclude `⌐◧-◧` from translations

solimander commented on packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css:
```diff
@@ -0,0 +1,70 @@
+.desktopLanguageButton {
```
I don't believe this file is used. If wallet styles are sufficient then generalize those for use here as well.

bcjgit replied:
It's used here:
https://sourcegraph.com/github.com/nounsDAO/nouns-monorepo@brianj_add_i18n_support_plus_ja_support/-/blob/packages/nouns-webapp/src/components/NavLocaleSwitcher/index.tsx?L128

I agree though that wallet dropdown styles are used in both places so will rename those to be more general

bcjgit replied:
done

solimander commented on packages/nouns-webapp/src/locales/en-US.js:
```diff
@@ -0,0 +1,317 @@
+/*eslint-disable*/ module.exports = {
```
The build catalogs typically should not be committed. Add to the `.gitignore` and compile in the `prepare` hook instead.

bcjgit commented:
@solimander update 5/1:

1. Refactored `NavWallet.module.css` into 2 classfiles (`NavWallet.module.css` and `NavBarDropdown.module.css`). The cross-cutting classes now being shared in the later per your comment. 
2.  Added build category to `.gitignore` and updated build to generate them on the fly accordingly
3. Excluded noggles from translation 
4. Updated error copy logic in `Bid.tsx` to update the `||` operator per your comment
5. Updated responsiveness logic 
6. Worked with translator to make a few minor changes to copy based on seeing them in the full context of the page
7. Made a handful of UI tweaks for things don't look strange in `ja_JP`

solimander commented on packages/nouns-webapp/package.json:
```diff
@@ -64,9 +69,12 @@
-    "eject": "react-scripts eject"
+    "eject": "react-scripts eject",
+    "i18n:extract": "lingui extract",
+    "i18n:compile": "lingui compile",
```
Add `"prepare": "yarn i18n:compile",` so localization works out of the box after installing dependencies for local development.

solimander commented:
nit: run `yarn format`

solimander reviewed (approved):
🇯🇵

KSST88 commented:
> # TLDR
> This PR does the following:
> 
> 1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
> 2. Add UI (mobile and desktop) to allow users to switch between Languages
> 3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
> 4. Add logic to detect user's system / browser locale (again, based off Uniswap)
> 5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)
> 
> ## Testing Done
> 1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
> 2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome)
> 3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic

KSST88 commented:
> # TLDR
> This PR does the following:
> 
> 1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
> 2. Add UI (mobile and desktop) to allow users to switch between Languages
> 3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
> 4. Add logic to detect user's system / browser locale (again, based off Uniswap)
> 5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)
> 
> ## Testing Done
> 1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
> 2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome)
> 3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic
